README.md
<p align="center">
  <a href="https://docs.logic-flow.cn" target="_blank">
    <img
      src="https://docs.logic-flow.cn/docs/_images/logo.png"
      alt="LogicFlow logo"
      width="250"
    />
  </a>
</p>

<p align="center">
  <a href="https://www.npmjs.com/package/chartflow-editor-core">
    <img src="https://img.shields.io/npm/v/chartflow-editor-core" alt="Version">
  </a>
  <a href="https://www.npmjs.com/package/chartflow-editor-core">
    <img src="https://img.shields.io/npm/dm/chartflow-editor-core" alt="Download">
  </a>
  <a href="https://github.com/didi/LogicFlow/blob/master/LICENSE">
    <img src="https://img.shields.io/npm/l/chartflow-editor-core" alt="LICENSE">
  </a>
</p>


简体中文 | [English](/README.en-us.md)

LogicFlow 是一款流程图编辑框架，提供了一系列流程图交互、编辑所必需的功能和简单灵活的节点自定义、插件等拓展机制，方便我们快速在业务系统内满足类流程图的需求。

## 特性


- 🛠 高拓展性
  
  兼容各种产品自定义的流程编辑需求，绝大部分模块以插件的形式实现，支持各模块自由插拔。

- 🚀 重执行
  
  流程图能完备的表达业务逻辑，不受流程引擎限制。

- 🎯 专业
  
  专注于业务流程图编辑的框架

## 使用

### 安装

```sh
# npm
$ npm install chartflow-editor-core chartflow-editor-extension --save

```

### 代码示例

```js
// 创建容器
<div id="container"></div>;

// 准备数据
const data = {
  // 节点
  nodes: [
    {
      id: '21',
      type: 'rect',
      x: 100,
      y: 200,
      text: '矩形节点',
    },
    {
      id: '50',
      type: 'circle',
      x: 300,
      y: 400,
      text: '圆形节点',
    },
  ],
  // 边
  edges: [
    {
      type: 'polyline',
      sourceNodeId: '50',
      targetNodeId: '21',
    },
  ],
};
// 渲染画布
const lf = new LogicFlow({
  container: document.querySelector('#container'),
  width: 700,
  height: 600,
});

lf.render(data);
```

## 文档

[官方文档](https://docs.logic-flow.cn)

- [快速上手](https://docs.logic-flow.cn/docs/#/zh/guide/start)
- [示例](https://docs.logic-flow.cn/examples/#/gallery)

非官方

- [Vuejs3.x + Typescript 上手指南](https://www.yuque.com/fe-xiaoxin/zeots5)
- [Vuejs3.x + Typescript 在线案例](https://1024code.com/collections/25)

## 核心能力

### 流程图编辑器快速搭建

提供了一个流程图编辑所必需的各项能力，这也是 LogicFlow 的基础能力：

- 图的绘制能力。基于 SVG 来绘制形状各异的节点和线，并提供了基础的节点（矩形、圆形、多边形等）和线（直线、折线、曲线）
- 各类交互能力，让图动起来。根据节点、线、图的各类鼠标事件（hover、点击、拖拽等）做出反应。比如节点拖拽、拖拽创建边、线的调整、双击节点编辑文本等
- 提升编辑效率的能力。提供网格、对齐线，上一步、下一步，键盘快捷键，图放大缩小等配套能力，帮助用户提升编辑效率
- 提供了丰富的 API ，宿主研发通过 API 传参调用和监听事件的方式，与 LogicFlow 完成交互

  下面是通过 LogicFlow 内置的节点和配套能力，做的流程图示例
  ：

    <image src="https://dpubstatic.udache.com/static/dpubimg/eEMT14E7BR/lfexample1.gif" width="500"/>

### 基于业务场景拓展

当基础能力无法满足业务需求的时候，便需要基于业务场景拓展。

- 设置图上所有元素的样式，比如各种节点、线、锚点、箭头、对齐线的大小颜色等，满足对前端样式调整的需求
- API 拓展。支持在 LogicFlow 上注册自定义的方法，比如通过 API 拓展提供图片下载的方法
- 自定义节点、线。内置的矩形、圆形等图形类节点往往无法满足实际的业务需求，需要定义具有业务意义的节点。LogicFlow 提供了 的方式让用户定制具有自定义图形、业务数据的节点，比如流程审批场景中的 “审批” 节点
- 拓展组件。LogicFlow 在 SVG 图层上提供了 HTML 层和一系列坐标转换逻辑，并支持在 HTML 层注册组件。宿主研发可以通过 LogicFlow 的 API，基于任何 View 框架开发组件，比如节点的右键菜单、控制面板等
- 数据转换 adapter。LogicFlow 默认导出的图数据不一定适合所有业务，此时可以通过 adapter API，在图数据从 LogicFlow 输入、输出的时候做自定义转换，比如转换成 BPMN 规范的图数据
- 内置部分拓展能力。基于上述拓展能力，我们还单独提供了 extension 的包，用来存放当前业务下沉淀出的具有通用性的节点、组件等，比如面向 BPMN 规范的节点和数据 adapter，默认菜单。注意 extension 可以单独安装，并支持按需引入

基于上述拓展的能力，前端研发能够根据实际业务场景的需求，灵活的开发出所需的节点、组件等。下面有两个基于 LogicFlow 拓展能力做出的流程图：

### BPMN应用demo

示例地址：http://logic-flow.org/examples/#/extension/bpmn

源码地址：https://github.com/didi/LogicFlow/tree/master/examples/src/pages/usage/bpmn

![图片:bpmn](https://dpubstatic.udache.com/static/dpubimg/CS6S6q9Yxf/lfexample2.gif)


#### 审批流应用demo

示例地址：http://logic-flow.org/examples/#/usage/approve

源码地址：https://github.com/didi/LogicFlow/tree/master/examples/src/pages/usage/approve

![图片: 审批流](https://dpubstatic.udache.com/static/dpubimg/uBeSlMEytL/lfexample3.gif)

#### vue 应用 demo

源码地址 [https://github.com/xinxin93/logicflow_vue_demo](https://github.com/xinxin93/logicflow_vue_demo)

![https://dpubstatic.udache.com/static/dpubimg/e35cef10-bb7c-4662-a494-f5aac024c092.gif](https://dpubstatic.udache.com/static/dpubimg/e35cef10-bb7c-4662-a494-f5aac024c092.gif)


#### 作图工具示例

LogicFlow不仅支持开发类似bpmn.js这种固定整体样式、更偏向生成数据在流程引擎可执行的流程图工具。也支持实现类似ProcessOn这种自由控制样式的作图工具。

示例地址：[http://logic-flow.org/mvp/index.html](http://logic-flow.org/mvp/index.html)

源码地址：[https://github.com/didi/LogicFlow/tree/master/site/mvp](https://github.com/didi/LogicFlow/tree/master/site/mvp)

示例图如下:
![logicflow-1.0-4.png](/docs/assets/images/LogicFlow-1.0-4.png)


#### vue3 node-red风格示例

源码地址: [https://github.com/Logic-Flow/logicflow-node-red-vue3](https://github.com/Logic-Flow/logicflow-node-red-vue3)

![node-red](https://cdn.jsdelivr.net/gh/Logic-Flow/static@latest/core/node-red.png)

#### 手绘风格插件示例

源码地址: [https://github.com/towersxu/draft-flow](https://github.com/towersxu/draft-flow)

![draft-flow](https://cdn.jsdelivr.net/gh/towersxu/draft-flow@latest/packages/website/public/redis.png)

## 联系我们

### 加入微信群

添加微信号：logic-flow 加入用户群

### 加入 QQ 群

<image src="https://dpubstatic.udache.com/static/dpubimg/VMBzV7jhh8/qq.png" width="300"/>

### 贡献代码

LogicFlow对外持开放态度，不论是只修改文档的标点符号还是对LogicFlow的整体功能有大的重构我们都欢迎。对于您的每一个PR我们都会认真查看、回复、合并。详情见[LogicFlow贡献指南](https://github.com/didi/LogicFlow/blob/master/CONTRIBUTING.md)。


examples/CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.6](https://github.com/mcknx/ChartFlowEditor/compare/examples@1.2.5...examples@1.2.6) (2023-03-30)

**Note:** Version bump only for package examples





## [1.2.5](https://github.com/mcknx/ChartFlowEditor/compare/examples@1.2.3...examples@1.2.5) (2023-03-29)

**Note:** Version bump only for package examples





## [1.2.4](https://github.com/mcknx/ChartFlowEditor/compare/examples@1.2.3...examples@1.2.4) (2023-03-29)

**Note:** Version bump only for package examples






## [1.2.3](https://github.com/didi/LogicFlow/compare/examples@1.2.2...examples@1.2.3) (2023-03-25)

**Note:** Version bump only for package examples





## [1.2.2](https://github.com/didi/LogicFlow/compare/examples@1.2.0...examples@1.2.2) (2023-03-25)

**Note:** Version bump only for package examples





## [1.2.1](https://github.com/didi/LogicFlow/compare/examples@1.2.0...examples@1.2.1) (2023-03-14)

**Note:** Version bump only for package examples





# [1.2.0](https://github.com/didi/LogicFlow/compare/examples@1.2.0-next.5...examples@1.2.0) (2023-03-13)

**Note:** Version bump only for package examples





# [1.2.0-next.5](https://github.com/didi/LogicFlow/compare/examples@1.2.0-next.4...examples@1.2.0-next.5) (2023-03-10)

**Note:** Version bump only for package examples





# [1.2.0-next.4](https://github.com/didi/LogicFlow/compare/examples@1.2.0-next.3...examples@1.2.0-next.4) (2023-02-22)

**Note:** Version bump only for package examples





# [1.2.0-next.3](https://github.com/didi/LogicFlow/compare/examples@1.2.0-next.2...examples@1.2.0-next.3) (2023-02-13)

**Note:** Version bump only for package examples





# [1.2.0-next.2](https://github.com/didi/LogicFlow/compare/examples@1.1.31...examples@1.2.0-next.2) (2023-01-13)


### Bug Fixes

* curved edge example ([20c953b](https://github.com/didi/LogicFlow/commit/20c953bc214583deacd7ce929047f17a00be784f))


### Features

* add event to dnd-penal ([9707cfb](https://github.com/didi/LogicFlow/commit/9707cfb927f13b0e6dd022230c5466f145d48cf8))





## [1.1.31](https://github.com/didi/LogicFlow/compare/examples@1.1.30...examples@1.1.31) (2022-11-12)
# [1.2.0-next.1](https://github.com/didi/LogicFlow/compare/examples@1.2.0-next.0...examples@1.2.0-next.1) (2022-10-10)

**Note:** Version bump only for package examples





## [1.1.30](https://github.com/didi/LogicFlow/compare/examples@1.1.29...examples@1.1.30) (2022-10-30)
# [1.2.0-next.0](https://github.com/didi/LogicFlow/compare/examples@1.1.29...examples@1.2.0-next.0) (2022-09-29)

**Note:** Version bump only for package examples





## [1.1.29](https://github.com/didi/LogicFlow/compare/examples@1.1.28...examples@1.1.29) (2022-09-24)


### Bug Fixes

* Fix typo: currrentProperties -> currentProperties ([71650b7](https://github.com/didi/LogicFlow/commit/71650b7c51d1ffd9c8b828f26ad2189e3b9cb7a4))





## [1.1.28](https://github.com/didi/LogicFlow/compare/examples@1.1.27...examples@1.1.28) (2022-09-17)

**Note:** Version bump only for package examples





## [1.1.27](https://github.com/didi/LogicFlow/compare/examples@1.1.26...examples@1.1.27) (2022-09-13)

**Note:** Version bump only for package examples





## [1.1.26](https://github.com/didi/LogicFlow/compare/examples@1.1.25...examples@1.1.26) (2022-08-27)

**Note:** Version bump only for package examples





## [1.1.25](https://github.com/didi/LogicFlow/compare/examples@1.1.24...examples@1.1.25) (2022-08-21)

**Note:** Version bump only for package examples





## [1.1.24](https://github.com/didi/LogicFlow/compare/examples@1.1.23...examples@1.1.24) (2022-08-11)

**Note:** Version bump only for package examples





## [1.1.23](https://github.com/didi/LogicFlow/compare/examples@1.1.22...examples@1.1.23) (2022-08-04)


### Bug Fixes

* 🐛 examples readme ([d1497a6](https://github.com/didi/LogicFlow/commit/d1497a65a61fe5742e677da2c005535581ffccaf))





## [1.1.22](https://github.com/didi/LogicFlow/compare/examples@1.1.21...examples@1.1.22) (2022-07-13)

**Note:** Version bump only for package examples





## [1.1.21](https://github.com/didi/LogicFlow/compare/examples@1.1.20...examples@1.1.21) (2022-07-02)

**Note:** Version bump only for package examples





## [1.1.20](https://github.com/didi/LogicFlow/compare/examples@1.1.19...examples@1.1.20) (2022-06-08)

**Note:** Version bump only for package examples





## [1.1.19](https://github.com/didi/LogicFlow/compare/examples@1.1.18...examples@1.1.19) (2022-06-01)

**Note:** Version bump only for package examples





## [1.1.18](https://github.com/didi/LogicFlow/compare/examples@1.1.17...examples@1.1.18) (2022-05-23)

**Note:** Version bump only for package examples





## [1.1.17](https://github.com/didi/LogicFlow/compare/examples@1.1.16...examples@1.1.17) (2022-05-23)

**Note:** Version bump only for package examples





## [1.1.16](https://github.com/didi/LogicFlow/compare/examples@1.1.14...examples@1.1.16) (2022-05-18)

**Note:** Version bump only for package examples





## [1.1.15](https://github.com/didi/LogicFlow/compare/examples@1.1.14...examples@1.1.15) (2022-05-07)

**Note:** Version bump only for package examples





## [1.1.14](https://github.com/didi/LogicFlow/compare/examples@1.1.13...examples@1.1.14) (2022-04-22)

**Note:** Version bump only for package examples





## [1.1.13](https://github.com/didi/LogicFlow/compare/examples@1.1.12...examples@1.1.13) (2022-04-16)

**Note:** Version bump only for package examples





## [1.1.12](https://github.com/didi/LogicFlow/compare/examples@1.1.12-alpha.0...examples@1.1.12) (2022-04-13)

**Note:** Version bump only for package examples





## [1.1.12-alpha.0](https://github.com/didi/LogicFlow/compare/examples@1.1.11...examples@1.1.12-alpha.0) (2022-04-09)

**Note:** Version bump only for package examples






examples/src/pages/usage/bpmn/config.js

export const approveUser = [
    {
      label: '直接上级',
      value: 'leader'
    },
    {
      label: 'T3领导',
      value: 't3Leader'
    },
    {
      label: 'T2领导',
      value: 't2Leader'
    },
    {
      label: 'T1领导',
      value: 't1Leader'
    },
  ]
  
  // 主题
  export const themeApprove = {
    rect: { // 矩形样式
      radius: 8,
    },
    circle: {
      r: 25,
    },
    // polyline: {
    //   strokeWidth: 1,
    // },
    edgeText: {
      background: {
        fill: 'white',
      },
    },
  }
  export const nodeType = {
    apply: '申请',
    approver: '审核',
    jugement: '判断',
    finsh: '结束'
  }
  
  export const data = {
    "nodes": [
      {
        "id": "28df2fbe-f32b-4a9b-b544-7e70d7187b33",
        "type": "apply",
        "x": 210,
        "y": 210,
        "text": { "x": 210, "y": 210, "value": "张三" },
        "properties": {
          "usernameZh": '张三',
          "username": 'zhangsan',
          "time": '2020-12-01 10:23:34',
          "result": '',
          "desc": '活动费用',
          "status": 1,
        }
      },
      {
        "id": "64179bd7-c60e-433c-8df7-97c8e98f855d",
        "type": "approver",
        "x": 350,
        "y": 210,
        "text": { "x": 350, "y": 210, "value": "李四" },
        "properties": {
          "labelColor": "#000000",
          "approveTypeLabel": "直接上级",
          "approveType": "leader",
          "usernameZh": '李四',
          "username": 'zhangsan',
          "time": '2020-12-01 11:23:34',
          "result": '通过',
          "desc": '',
          "status": 1,
        }
      },
      {
        "id": "fcb96f10-720e-40e5-8ed0-ebdd0a46f234",
        "type": "jugement",
        "x": 510,
        "y": 210,
        "text": { "x": 510, "y": 210, "value": "判断报销是否\n大于1000元" },
        "properties": {
          "api": "",
          "status": 1,
          "usenameZh": "系统",
          "desc": '判断报销是否大于1000元'
        },
      },
      {
        "id": "9f119df3-c449-4e5d-a67a-cb351b9cbdb5",
        "type": "approver",
        "x": 670,
        "y": 210,
        "text": { "x": 670, "y": 210, "value": "王五" },
        "properties": {
          "labelColor": "#000000",
          "approveTypeLabel": "T2领导",
          "approveType": "t2Leader",
          "usernameZh": '王五',
          "username": 'wangwu',
          "time": '2020-12-01 12:23:34',
          "result": '驳回',
          "desc": '',
          "status": 2,
        }
      },
      {
        "id": "ef34f09c-38ea-4ad4-acd2-cc2f464a2be6",
        "type": "finsh",
        "x": 850,
        "y": 210,
        "text": { "x": 850, "y": 210, "value": "结束" },
        "properties": {
          "status": 2,
        },
      }
    ],
    "edges": [
      {
        "id": "0d87b1eb-2389-445a-9f34-6227940b2072",
        "type": "action",
        "sourceNodeId": "28df2fbe-f32b-4a9b-b544-7e70d7187b33",
        "targetNodeId": "64179bd7-c60e-433c-8df7-97c8e98f855d",
        "startPoint": { "x": 235, "y": 210 },
        "endPoint": { "x": 300, "y": 210 },
        "text": { "x": 51.25, "y": 0, "value": "" },
        "properties": {
          "status": 1,
        },
        "pointsList": [
          { "x": 235, "y": 210 },
          { "x": 300, "y": 210 }
        ]
      },
      {
        "id": "d99e7451-b379-411e-b0da-df11be8be20a",
        "type": "action",
        "sourceNodeId": "64179bd7-c60e-433c-8df7-97c8e98f855d",
        "targetNodeId": "fcb96f10-720e-40e5-8ed0-ebdd0a46f234",
        "startPoint": { "x": 400, "y": 210 },
        "endPoint": { "x": 475, "y": 210 },
        "text": { "x": 437.5, "y": 210, "value": "通过" },
        "properties": {
          "status": 1,
        },
        "pointsList": [
          { "x": 400, "y": 210 },
          { "x": 475, "y": 210 }
        ]
      },
      {
        "id": "4c615802-15d8-442c-be22-b65430286123",
        "type": "action",
        "sourceNodeId": "fcb96f10-720e-40e5-8ed0-ebdd0a46f234",
        "targetNodeId": "9f119df3-c449-4e5d-a67a-cb351b9cbdb5",
        "startPoint": { "x": 545, "y": 210 },
        "endPoint": { "x": 620, "y": 210 },
        "text": { "x": 582.5, "y": 210, "value": "是" },
        "properties": {
          "status": 1,
        },
        "pointsList": [
          { "x": 545, "y": 210 },
          { "x": 620, "y": 210 }
        ]
      },
      {
        "id": "934ae03a-6ee0-4568-a2b4-8bcede565e0b",
        "type": "action",
        "sourceNodeId": "9f119df3-c449-4e5d-a67a-cb351b9cbdb5",
        "targetNodeId": "ef34f09c-38ea-4ad4-acd2-cc2f464a2be6",
        "startPoint": { "x": 720, "y": 210 },
        "endPoint": { "x": 825, "y": 210 },
        "text": { "x": -10, "y": 0, "value": "驳回" },
        "properties": {
          "status": 2,
        },
        "pointsList": [
          { "x": 720, "y": 210 },
          { "x": 825, "y": 210 }
        ]
      },
    ]
  }
  
  